### PR TITLE
feat(TG-246, TG-250): Facebook auth with passport-facebook-token and merge FB account

### PR DESCRIPTION
**Pull request by sergey-trotsyuk**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/garage-it/SmartHouse-backend/pull/74" target="_blank">ReviewNinja

File: src/API/auth/facebook/passport.js [hunk 1]
```diff
@@ -1,28 +1,41 @@
...
-        })
+        if (profile.emails[0].value) {
+            // Some times it happens
+            // https://developers.facebook.com/bugs/298946933534016/?comment_id=620710881344692
+            // https://developers.facebook.com/docs/graph-api/reference/user
+            done(new Error('Facebook Auth: Not found e-mail in profile'));
+            return;
+        }
+
+        UserService.findOrCreateUser(
```
Comment (borys-ihnatyev) [inline src/API/auth/facebook/passport.js]: Is it possible to pass only the profile as argument? Or there is other use cases expected ?

File: src/shared/user/user.service.js [hunk 2]
```diff
@@ -51,15 +51,27 @@ function createUser(body) {
...
  * @function findOrCreateUser
  * @description It will find user by query or create new user in database
  * @param {object} query
  * @param {object} body
  * @return {object}
  */
-function findOrCreateUser (query, body) {
-    return User.findOrCreate(query, body);
+function findOrCreateUser(query, body) {
+    return new Promise((resolve, reject) => {
```
Comment (borys-ihnatyev) [inline src/shared/user/user.service.js]: I saw that we use ```bluebird``` as promise library

File: src/shared/user/user.service.js [hunk 3]
```diff
@@ -51,15 +51,27 @@ function createUser(body) {
...
+        User.findOrCreate(query, body, {}, (err, user) => {
+            if (err) {
```
Comment (borys-ihnatyev) [inline src/shared/user/user.service.js]: this is redundant, ```User.findOrCreate``` will return promise if you did not pass ```function``` as the last argument

File: src/API/auth/auth.controller.js [hunk 4]
```diff
@@ -1,20 +1,27 @@
...
 
+import config from '../../config/env';
+
 /**
  * @function initialize
  * @description Register authenticate strategies
  */
 function initialize() {
     localSetup(UserService);
+    facebookSetup(UserService, config.facebook);
```
Comment (borys-ihnatyev) [inline src/API/auth/auth.controller.js]: inner functions sufficed with ```Setup```, why not name the caller ```setup```

File: src/API/auth/auth.controller.js [hunk 5]
```diff
@@ -37,34 +44,46 @@ function register() {
...
-                token: authService.generateToken(req.user._id)
-            });
-        }
+        bodyToParamsMiddleware,
+        createPassportAuthenticateMiddleware('local'),
+        generateTokenControllerAction
+    );
+}
+
+function facebookLoginWithAccessToken() {
```
Comment (borys-ihnatyev) [inline src/API/auth/auth.controller.js]: I think that ```WithAccessToken``` suffix is redundant, in the reason that we have no opposite method called ```facebookLoginWithoutAccessToken```

File: src/shared/user/user.service.js (hunk 2 above)
Comment (sergey-trotsyuk) [inline src/shared/user/user.service.js]: Will fix, thank you.

File: src/API/auth/auth.controller.js (hunk 5 above)
Comment (sergey-trotsyuk) [inline src/API/auth/auth.controller.js]: Facebook auth can be implemented with "token" strategy and with "code" strategy, so this name highlights strategy.

File: src/API/auth/auth.controller.js (hunk 4 above)
Comment (sergey-trotsyuk) [inline src/API/auth/auth.controller.js]: Because we have localSetup and facebookSetup.

File: src/API/auth/facebook/passport.js [hunk 6]
```diff
@@ -1,28 +1,41 @@
...
  * @param {Object} UserService
- * @param {Object} config
+ * @param {Object} facebookConfig
  */
-export default function setup(UserService, config) {
-    passport.use(new FacebookStrategy({
-        clientID: config.facebook.clientID,
-        clientSecret: config.facebook.clientSecret,
-        callbackURL: config.facebook.callbackURL
+export default function setup(UserService, facebookConfig) {
```
Comment (sergey-trotsyuk) [inline src/API/auth/facebook/passport.js]: Yes, but I think injection is create some borders :). And it should help when we start write unit tests.

File: src/API/auth/facebook/passport.js [hunk 7]
```diff
@@ -1,28 +1,41 @@
...
+    passport.use(new FacebookTokenStrategy({
+        clientID: facebookConfig.clientID,
+        clientSecret: facebookConfig.clientSecret
     }, (accessToken, refreshToken, profile, done) => {
-        UserService.findOrCreateUser({ email: profile.emails[0].value }, {
-            facebookId: profile.id,
-            email: profile.emails[0].value,
-            name: profile.username
-        })
+        if (profile.emails[0].value) {
```
Comment (yakimko) [inline src/API/auth/facebook/passport.js]: please move profile.emails[0].value to const

Comment (yakimko) [inline src/API/auth/facebook/passport.js]: And maybe 
```
if (!profile.emails[0].value)
```

Comment (sergey-trotsyuk) [inline src/API/auth/facebook/passport.js]: Fixed, thank you.

Comment (sergey-trotsyuk) [inline src/API/auth/facebook/passport.js]: Fixed

**Pull request merged**